opm/autodiff/BlackoilAmg.hpp

```diff
@@ -89,30 +90,35 @@ Dune::OverlappingSchwarzOperator<M,X,Y,T> createOperator(const Dune::Overlapping
 //! \param comm The communication objecte describing the data distribution.
 //! \param pressureIndex The index of the pressure in the matrix block
 //! \retun A pair of the scaled matrix and the associated operator-
-template<class Operator, class Communication>
+template<class Operator, class Communication,class Vector>
 std::tuple<std::unique_ptr<typename Operator::matrix_type>, Operator>
-scaleMatrixQuasiImpes(const Operator& op, const Communication& comm,
-                      std::size_t pressureIndex)
+scaleMatrixDRS(const Operator& op, const Communication& comm,
+	       std::size_t pressureIndex,const Vector& weights, const Opm::CPRParameter& param)
 {
     using Matrix = typename Operator::matrix_type;
     using Block = typename Matrix::block_type;
+    using BlockVector = typename Vector::block_type;
     std::unique_ptr<Matrix> matrix(new Matrix(op.getmat()));
-
-    for ( auto& row : *matrix )
-    {
-        for ( auto& block : row )
-        {
-            for ( std::size_t i = 0; i < Block::rows; i++ )
-            {
-                if ( i != pressureIndex )
-                {
-                    for(std::size_t j=0; j < Block::cols; j++)
-                    {
-                        block[pressureIndex][j] += block[i][j];
-                    }
-                }
-            }
-        }
+    if(param.cpr_use_drs_){
+      const auto endi = matrix->end();
+      for (auto i=matrix->begin(); i!=endi; ++i){
+	const BlockVector& bw = weights[i.index()];
+	const auto endj = (*i).end();
+	  for (auto j=(*i).begin(); j!=endj; ++j){  
+	    {
+	      BlockVector bvec(0.0);
```

Comment:
Probably a matter of taste but I find this loop hard to read. I think you are handcoding
```
BlockVector bvec(0.0);
block.mtv(bw, block);
block[pressureIndex] = bvec;
```